Revamp go-ipfs Release Flow (#6482)
ipfs/go-ipfs: docs/releases.md

# `go-ipfs` Release Flow

## Table of Contents

- [Release Philosophy](#release-philosophy)
- [Release Flow](#release-flow)
- [Performing a Release](#performing-a-release)
- [Release Version Numbers (aka semver)](#release-version-numbers-aka-semver)

## Release Philosophy

> `go-ipfs` aims to have release every six weeks, two releases per quarter. During these 6 week releases, we go through 4 different stages of Release Candidates (RC) that gives us the opportunity to test the new version against our test envinronments, production envinronments, IPFS apps (e.g. Desktop and WebUI) and with our close partners that have IPFS running in Production, by leveraging their own test infrastructure.

Reviewer: > close partners

This terminology gives a sense of gatekeeping or a private test programme, e.g. what are the criteria to be considered a _close partner_ in the eyes of the open community? Who decides who receives that qualification, and how? I'd advocate to shift the terminology here and establish a transparent process:

1. Maybe call this "early testers programme"?
2. Maintain an `EARLY_TESTERS.md` file which is a table of Github handles to be notified when a release is available to be tested.
3. Addition and removal is by self-selection via PR on that file. We add those people to a Github team, so we can notify them at once.
4. When we produce a new release and want to subject it to early testing, we open an issue with the release notes, and mention the group.
5. Folks report their findings back on that issue, or by opening new issues referencing the _early testing_ issue.
6. We can pin those _early testing_ issues for visibility.

---

nit: typo in 'envinronments' x2

> `go-ipfs` aims to have release every six weeks, two releases per quarter. During these 6 week releases, we go through 4 different stages of Release Candidates (RC) that gives us the opportunity to test the new version against our test envinronments, production envinronments, IPFS apps (e.g. Desktop and WebUI) and with our close partners that have IPFS running in Production, by leveraging their own test infrastructure.

Reviewer: > close partners

This terminology gives a sense of gatekeeping or a private test programme, e.g. what are the criteria to be considered a _close partner_ in the eyes of the open community? Who decides who receives that qualification, and how? I'd advocate to shift the terminology here and establish a transparent process:

1. Maybe call this "early testers programme"?
2. Maintain an `EARLY_TESTERS.md` file which is a table of Github handles to be notified when a release is available to be tested.
3. Addition and removal is by self-selection via PR on that file. We add those people to a Github team, so we can notify them at once.
4. When we produce a new release and want to subject it to early testing, we open an issue with the release notes, and mention the group.
5. Folks report their findings back on that issue, or by opening new issues referencing the _early testing_ issue.
6. We can pin those _early testing_ issues for visibility.

---

nit: typo in 'envinronments' x2

Author: Big thumbs up for the rename and making it clear of what are the steps to join that program

Reviewer 2: We also could have a mailing list for pre-releases and final releases. This would be useful for package maintainers that would like to be notified about the release so they can update package in a distribution.

We might expand the six week release schedule in case of:
- No new updates to be added
- In case of a large community event that takes the core team availability away (e.g. IPFS Conf, Dev Meetings, IPFS Camp, etc)

## Release Flow

`go-ipfs` releases come in 4 stages:

- **Release Stage 1 - Internal testing** - Test the Release against our testing infrastructure, including interoperability, integration, test lab, multiple runtimes and the apps we've built (WebUI, Desktop, NPM on IPFS, HTTP Client Libraries). The intent is to make this stage fully automated (and somewhat is already), until then, we manually check a list and ensure that all tests have been run

> - **Release Stage 2 - Invite IPFS partners to test** - Reach out to our partners (i.e. projects that have volunteered to support `go-ipfs` by using their own test infrastructure and tell us the results)

Reviewer: Drawing from the above comments, I'd suggest changing the word _partners_ to _early testers_.

> - **Release Stage 3 - Announce to the broader community** - Communicate to the community that a new Release Candidate is ready and that everyone is welcome to test it with us

Reviewer 3: Any real reason not to combine stages 2 and 3? Really, we usually ask the broader community (stage 3) first because they tend not to be using IPFS for mission critical operations. That is:

1. Advanced/eager users tend to find "shallow" bugs: api breakage, bad documentation, etc.
2. Partners tend to find deep bugs: performance regressions, rare crashes, etc.

Author: My rationale for this separation was based on my assumption that we are very limited in capacity to deal with incoming bug reports.

I expect that our early testers are more knowledgeable and therefore, more capable of debugging themselves part of the problem or at least providing more useful information, which will save us a ton of time and cover a ton of bugs. 

In contrast, I expect that the broad community to be able to also pick on bugs but provide less useful information. By doing early testers first, we save some time from debugging less documented/debugged bugs (which would overlap with the ones that our early testers will report).

Author: That said, I'm not holding any strong feelings towards the flow. Happy to take in the wisdom of the maintainers and iterate on it :)

Reviewer 3: What about:

* Stage 2 - Announce a preliminary RC on IRC and invite the community to run sanity checks.
* Stage 3 - Pull in partners.
* Stage 4 - Announce the RC on twitter etc? I wonder if there's some way to allow, e.g., ipfs-desktop users to opt into this release.

I guess I mostly just want a chance to test it with _lower_ stakes than people relying on IPFS in production systems.

> My rationale for this separation was based on my assumption that we are very limited in capacity to deal with incoming bug reports.

Most of the bug reports we get from announcements on IRC are actually pretty good. They also tend to be pretty straight forward to diagnose as they aren't subtle performance issues.

Author: > lower stakes than people relying on IPFS in production systems.

We should not advise folks to run RC in production, rather just invite them to test on their test infra and QA if they have it built.

> Most of the bug reports we get from announcements on IRC are actually pretty good. They also tend to be pretty straight forward to diagnose as they aren't subtle performance issues.

This is good info to know. If you feel confident on merging Stage 2 and 3, then let's do it. Just confirm and I'll update it

Reviewer 3: > We should not advise folks to run RC in production, rather just invite them to test on their test infra and QA if they have it built.

We _need_ to run our final RC in production before releasing it. Basically, we have two options:

1. We cut a release with a performance regression an wait a release cycle (or go back through the release cycle for a patch release).
2. We cut release _candidate_ and ask partners to start running it. This way, we can fix any performance regressions without starting a new release cycle.

This RC should be a true RC, a candidate for release.

I believe this comes down to an issue of naming. Really, the stages are:

1. Alpha - Internal: Small sample size, light and heavy workloads.
2. Beta - IRC: Medium sample size, light workloads.
3. RC1 - Partners: Medium sample size, heavy workloads.
4. RC2 - End users (twitter): Large sample size.

However, IPFS itself is "alpha" so this would be even more confusing.

I'd _love_ to get partners on board at the "beta" stage but I'm not sure how helpful that will be unless they're willing to build out simulations for us. Having more testing on light workloads won't help all that much.

Reviewer 4: IMHO, RC1 (using @Stebalien terminology ^), should be run in production by more than just us. It doesn’t need to be nor should it be a full scale deployment of the RC1 version but to assume that deploying an RC1 to partner test/QA infra where the load is no where near what it is in production is going to lead us to believe things are fine when they are not. The burden that we place on our partners in this process should be as light as possible, and bringing them in at the RC1 stage is about ensuring that any changes made to go-ipfs do not lead to regressions in performance or functionality when running in their particular usecase/environment, which to me indicates that they should be able to deploy at least one instance of the RC1 to production without any dire consequences.

Reviewer 5: It's sounds weird that certain releases are meant for partners. It sounds like giving early access to features even though it means "try it out". It should be very clear that anyone can try it out. I can also imagine that different partners (and users) may choose to test ipfs pre-releasees at different stages depending on circumstances.

Note that it is also weird that we give some partners a half-baked release to test. Either they test it by deploying it to their users, in which case they should not be given a half-baked release that can break their apps (even if they volunteer), or they have they're own testing environments, in which case they could directly master or a release branch continuously without waiting for a specific release stage.

Otherwise said:

  * Sounds good to have alpha/beta/rcs (we should have a release branch that includes all features in the release and only updates with bugfixes). We should be able to skip certain checkpoints sometimes (like hotfixes to specific issues).
  * Document what the level of maturity is after each checkpoint, so that everyone can test and try it out (partner or not).
  * Create a known and consistent venue to gather feedback on each release (i.e. an issue) and include information from our own testing in it.
  * Let people choose when they test based on their availability, feedback on the current state of the release and issues with the previous release.

Author: **We absolutely must not encourage partners to run a RC in production that can potentially harm their service and hurt their users.** What we hope is for partners to use their own Beta channels, Testing Infrastructure and potentially QA setup to let us know if the new release passes the tests.

Reviewer 3: > We absolutely must not encourage partners to run a RC in production that can potentially harm their service and hurt their users.

Any release we make can potentially break something. This is an inviolate rule of software and we must not lie to ourselves about this. The goal here is to minimize the impact by deploying in stages, not to avoid all harm.

We definitely shouldn't ask our partners to run RCs that haven't been tested but we should absolutely ask them to run an RC that we believe to be "ready for release", to reduce the impact of any bugs they may encounter. That way they can roll back, we can fix the issue, and then move on with the release cycle without waiting for another release.

> What we hope is for partners to use their own Beta channels, Testing Infrastructure and potentially QA setup to let us know if the new release passes the tests.

This would be awesome but it's asking a lot from our partners and isn't currently the weakest part of our release process.

Reviewer 3: You know, we probably _can_ just call these alpha, beta, etc. That should make this easier to explain to users:

* Dev - Bleeding Edge.
* Alpha - Feature Freeze.
* Beta - Tested but may still be buggy.
* RC1 - Release quality, small rollout.
* RC2 - Release quality, larger rollout.
* Release - Full rollout.

Reviewer 3: See: https://github.com/ipfs/go-ipfs/pull/6496

- **Release Stage 4 - Complete the Release** - Finalize the release, start the next release.

The Release Stages are not linked to Release Candidate numbers, in fact, there can be multiple release candidate per stages as we catch bugs and improve the release itself.

<p align="center">
  <a href="https://ipfs.io">
    <img src="https://gateway.ipfs.io/ipfs/QmaFtLxoCAm5vFQ9AftKkhJwSAdDdF1jzV9DfzW6gbXqFL/Paper.Sketches.23.png" width="450" />
  </a>
</p>

## Performing a Release

The first step is for the `Lead Maintainer` for `go-ipfs` to open an issue with Title `go-ipfs <version> Release` and a c&p of the following template:

```
> <short tl;dr; of the release>

# 🗺 What's left for release

<List of items with PRs and/or Issues to be considered for this release>

# 🔦 Highlights

> <List of items with PRs and/or Issues to be considered for this release>

Reviewer 3: ? You mean already included?

# 🏗 API Changes

> <List of API changes, if any>

Author: The `🔦 Highlights` and `🏗 API Changes` could potentially be replaced by the Changelog. That said, there is still value on highlighting some of the changes, the ones that we want users to know about first (or the ones that we want to ensure that users see, in case they don't have time to read the entire changelog).

Reviewer 3: Or, more generally, breaking changes/notices.

# ✅ Release Checklist

For each RC published in each stage:
- [ ] version string in `version.go` has been updated
- [ ] tag commit with vX.Y.Z-rcN

> ### Release Stage 1 - Internal testing

Author: Note: as noted on Release Flow, this whole stage can potentially be automated. Until it is, let's have the items as a check list so that we don't forget any :)

When Release Stage 1, there is a features freeze for the release branch.

- [ ] CHANGELOG.md has been updated
  - use `./bin/mkreleaselog` to generate a nice starter list
- [ ] Automated Testing - Ensure that all tests are passing, this includes:
  - [ ] unit
  - [ ] sharness
  - [ ] [interop](https://github.com/ipfs/interop#test-with-a-non-yet-released-version-of-go-ipfs)
- [ ] Network Testing:

>   - [ ] test lab things

Reviewer 3: We don't have this yet.

Author: @bigs isn't there a thing?

Reviewer 4: It is a thing but not yet for IPFS just yet.

- [ ] Infrastructure Testing:

>   - [ ] Deploy new version to a subset of Bootstrappers

Reviewer 3: Can we even do this yet? cc @ipfs/wg-infrastructure.

Author: @mburns ?

  - [ ] Deploy new version to a subset of Gateways
  - [ ] Deploy new version to a subset of Preload nodes
  - [ ] Collect metrics every day. Work with the Infrastructure team to learn of any hiccup
- [ ] IPFS HTTP Client Libraries Testing:

>   - [ ] [JS](http://github.com/ipfs/js-ipfs-http-client)

Reviewer 3: These are built into the interop tests.

  - [ ] [Go](https://github.com/ipfs/go-ipfs-api)
- [ ] IPFS Application Testing -  Run the tests of the following applications:
  - [ ] WebUI
  - [ ] IPFS Desktop
  - [ ] IPFS Companion
  - [ ] NPM on IPFS

### Release Stage 2 - Invite IPFS partners to test

- [ ] Reach out to the following IPFS partners for testing this release (check when no more problems have been reported):
  - [ ] Infura
  - [ ] Textila
  - [ ] Open Bazaar

>   - [ ] QRI

Reviewer 3: Let's add Pinata and remove Open Bazaar (forked version).

- [ ] Run tests available in the following repos with the latest RC (check when all tests pass):

>   - [ ] [orbit-db](https://github.com/orbitdb/orbit-db)

Author: I'm pretty sure that both of the lists here can be way bigger. We probably should have a "expectations description" before adding more folks

Reviewer 3: Can we test this against go-ipfs? I thought it embedded a js-ipfs node.

Author: AFAIK, it can do both. Will check again

Reviewer 3: (yes, sorry)

PSA: If you are a heavy user of `go-ipfs`, have developed a solid test infrastructure for your application and would love to help us would like to help us test `go-ipfs` release candidates, reach out to go-ipfs-wg@ipfs.io.

### Release Stage 3 - Announce to the broader community

- [ ] Documentation
  - [ ] Ensure that CHANGELOG.md is up to date
  - [ ] Ensure that README.md is up to date
  - [ ] Ensure that all the examples we have produced for go-ipfs run without problems
  - [ ] Update HTTP-API Documentation on the Website using https://github.com/ipfs/http-api-docs
- [ ] Invite the community through (link to the release issue):
  - [ ] [discuss.ipfs.io](https://discuss.ipfs.io/c/announcements)

>   - [ ] Twitter

Reviewer 3: And IRC?

### Release Stage 4 - Complete the Release

> - Take a snapshot between of everyone that has contributed to this release (including its subdeps in IPFS, libp2p, IPLD and multiformats) using [`name-your-contributors`](https://www.npmjs.com/package/name-your-contributors). Generate a nice markdown list with [this script](https://gist.github.com/alanshaw/5a2d9465c5a05b201d949551bdb1fcc3). Add it to this issue.

Author: This gist needs to be updated to point to the go code repos. Thanks to @alanshaw for creating it in the first place.

Reviewer 3: The `bin/mkreleaselog` script currently does this. Output: https://blog.ipfs.io/93-go-ipfs-0.4.21#huge-thank-you-to-everyone-that-made-this-release-possible

  - [ ] tag commit with vX.Y.Z
  - [ ] update release branch to point to release commit (`git merge vX.Y.Z`).
  - [ ] publish dist.ipfs.io
  - [ ] publish next version to https://github.com/ipfs/npm-go-ipfs
- [ ] Publish a Release Blog post (at minimum, a c&p of this release issue with all the highlights, API changes, link to changelog and thank yous)
- [ ] Broadcasting (link to blog post)
  - [ ] Twitter
  - [ ] IRC
  - [ ] Reddit
  - [ ] [discuss.ipfs.io](https://discuss.ipfs.io/c/announcements)
  - [ ] Announce it on the [IPFS Users mlist](https://groups.google.com/forum/#!forum/ipfs-users)

# ❤️ Huge thank you to everyone that made this release possible

In alphabetical order, here are all the humans that contributed to the release:

- <use script -- listed in Release Stage 4 -- to generate a list of everyone that contributed for this release>

# 🙌🏽 Want to contribute?

Would you like to contribute to the IPFS project and don't know how? Well, there are a few places you can get started:

- Check the issues with the `help wanted` label in the [go-ipfs repo](https://github.com/ipfs/go-ipfs/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22)

> - Join an IPFS All Hands, introduce yourself and let us know where you would like to contribute - https://github.com/ipfs/team-mgmt/#weekly-ipfs-all-hands

Reviewer 3: Maybe we should start pointing people at IRC/Matrix instead of the All Hands.

Author: Multiple people have reported that showing up on the main channel on IRC with ideas and questions typically results into them falling into oblivion.

If someone is new, it is nice to see that there are humans, that is why the call is important.

Reviewer 3: Fair enough.

- Hack with IPFS and show us what you made! The All Hands call is also the perfect venue for demos, join in and show us what you built
- Join the discussion at http://discuss.ipfs.io/ and help users finding their answers.

> - Join the [Go Core Dev Team Weekly Sync](https://github.com/ipfs/team-mgmt/issues/650) and be part of the Sprint action!

Reviewer 3: Gone for now.

Author: Is there a replacement?

Author: @momack2 ?

# ⁉️ Do you have questions?

> The best place to ask your questions about IPFS, how it works and what you can do with it is at [discuss.ipfs.io](http://discuss.ipfs.io). We are also available at the `#ipfs` channel on Freenode.

Reviewer 3: + matrix

```

## Release Version Numbers (aka semver)

> Until `go-ipfs` 0.4.X, `go-ipfs` was not using semver to communicate the type of relase

Reviewer 6: nit 
>release

Author: wanna do a comment with a commit fix?

Reviewer 6: github getting so fancy on me.

Author: Thank you :D

Post `go-ipfs` 0.5.X, `go-ipfs` will use semver. This means that patch releases will not contain any breaking changes nor new features. Minor releases might contain breaking changes and always contain some new feature

Post `go-ipfs` 1.X.X (future), `go-ipfs` will use semver. This means that only major releases will contain breaking changes, minors will be reserved for new features and patches for bug fixes.

> We do not yet retroactively apply fixes to older releases (no Long Term Support releases for now), which means that we always recommend users to update to the latest, whenever possible.

Reviewer 7: Ho, bummer...

Please consider this in the future for disruptive problems. At Infura we have been stuck with problematic release for months without solution (other than forking): 0.4.19 and 0.4.20 had memory usage problem but we couldn't revert to 0.4.18 due to security reason. This will only become more critical in the future.

Reviewer 3: In that case, we would have applied the CPU fixes in 0.4.19 to 0.4.18 as a patch. This is just saying that we're not going to release a 0.4.20 and then go back and patch a 0.4.18 release.

# `go-ipfs` Release Flow

## Table of Contents

- [Release Philosophy](#release-philosophy)
- [Release Flow](#release-flow)
- [Performing a Release](#performing-a-release)
- [Release Version Numbers (aka semver)](#release-version-numbers-aka-semver)

## Release Philosophy

> `go-ipfs` aims to have release every six weeks, two releases per quarter. During these 6 week releases, we go through 4 different stages of Release Candidates (RC) that gives us the opportunity to test the new version against our test environments (unit, interop, integration), QA in our current production environmment, IPFS apps (e.g. Desktop and WebUI) and with our _early testers_<sup>[1]</sup> that have IPFS running in Production, by leveraging their own test infrastructure and QA systems.

Reviewer 8: ```suggestion
`go-ipfs` aims to have release every six weeks, two releases per quarter. During these 6 week releases, we go through 4 different stages of Release Candidates (RC) that gives us the opportunity to test the new version against our test environments (unit, interop, integration), QA in our current production environment, IPFS apps (e.g. Desktop and WebUI) and with our _early testers_<sup>[1]</sup> that have IPFS running in Production, by leveraging their own test infrastructure and QA systems.
```

We might expand the six week release schedule in case of:
- No new updates to be added
- In case of a large community event that takes the core team availability away (e.g. IPFS Conf, Dev Meetings, IPFS Camp, etc)

## Release Flow

`go-ipfs` releases come in 4 stages:

- **Release Stage 1 - Internal testing** - Test the Release against our testing infrastructure, including interoperability, integration, test lab, multiple runtimes and the apps we've built (WebUI, Desktop, NPM on IPFS, HTTP Client Libraries). The intent is to make this stage fully automated (and somewhat is already), until then, we manually check a list and ensure that all tests have been run
- **Release Stage 2 - Invite _early testers_ to try it out** - Reach out to our _early testers_ (i.e. projects that have volunteered to support `go-ipfs` by using their own test infrastructure and tell us the results)
- **Release Stage 3 - Announce to the broader community** - Communicate to the community that a new Release Candidate is ready and that everyone is welcome to test it with us
- **Release Stage 4 - Complete the Release** - Finalize the release, start the next release.

> The Release Stages are not linked to Release Candidate numbers, in fact, there can be multiple release candidate per stages as we catch bugs and improve the release itself.

Author: The plan is just to use one number. The RC-<number> signals that there has been a new one published

> The Release Stages are not linked to Release Candidate numbers, in fact, there can be multiple release candidate per stages as we catch bugs and improve the release itself.

Reviewer 3: See my comment on alpha, beta, etc. above.

# 🔦 Highlights

<List of items with PRs and/or Issues included for this release>

# 🏗 API Changes

<List of API changes, if any>

# ✅ Release Checklist

For each RC published in each stage:
- [ ] version string in `version.go` has been updated
- [ ] tag commit with vX.Y.Z-rcN

### Release Stage 1 - Internal testing

When Release Stage 1, there is a features freeze for the release branch.

- [ ] CHANGELOG.md has been updated
  - use `./bin/mkreleaselog` to generate a nice starter list
- [ ] Automated Testing - Ensure that all tests are passing, this includes:
  - [ ] unit
  - [ ] sharness
  - [ ] [interop](https://github.com/ipfs/interop#test-with-a-non-yet-released-version-of-go-ipfs)
- [ ] Network Testing:
  - [ ] test lab things
- [ ] Infrastructure Testing:
  - [ ] Deploy new version to a subset of Bootstrappers
  - [ ] Deploy new version to a subset of Gateways
  - [ ] Deploy new version to a subset of Preload nodes
  - [ ] Collect metrics every day. Work with the Infrastructure team to learn of any hiccup
- [ ] IPFS HTTP Client Libraries Testing:
  - [ ] [JS](http://github.com/ipfs/js-ipfs-http-client)
  - [ ] [Go](https://github.com/ipfs/go-ipfs-api)

> - [ ] IPFS Application Testing -  Run the tests of the following applications:

Reviewer 4: IPFS Cluster?

Reviewer 3: We're going to need DRIs for this. Maybe add a line about pinging the relevant maintainers?

  - [ ] WebUI
  - [ ] IPFS Desktop
  - [ ] IPFS Companion
  - [ ] NPM on IPFS

### Release Stage 2 - Invite _early testers_ to try it out

- [ ] Reach out to the following IPFS _early testers_ for testing this release (check when no more problems have been reported):
  - [ ] Infura
  - [ ] Textila
  - [ ] Pinata
  - [ ] QRI
- [ ] Run tests available in the following repos with the latest RC (check when all tests pass):
  - [ ] [orbit-db](https://github.com/orbitdb/orbit-db)

PSA: If you are a heavy user of `go-ipfs`, have developed a solid test infrastructure for your application and would love to help us would like to help us test `go-ipfs` release candidates, reach out to go-ipfs-wg@ipfs.io.

### Release Stage 3 - Announce to the broader community

- [ ] Documentation
  - [ ] Ensure that CHANGELOG.md is up to date

>   - [ ] Ensure that README.md is up to date

Reviewer 8: I past few years we had times when parts of `docs/config.md` and `docs/experimental-features.md` got out of date. Its the source of truth for many power users. It should be part of checklist:
```suggestion
  - [ ] Ensure that README.md is up to date
  - [ ] Ensure that `docs/config.md` and `docs/experimental-features.md` are up to date
```

At this stage, we'll:
- 1. Start a partial-rollout to our own infrastructure.
- 2. Test against ipfs and ipfs-shipyard applications.

**Goal(s):**
- 1. Make sure we haven't introduced any obvious regressions.
- 2. Test the release in an environment we can monitor and easily roll back (i.e., our own infra).

### Stage 2 - Public Beta

At this stage, we'll announce the impending release to the community and ask for beta testers.

**Goal:** Test the release in as many non-production environments as possible. This is relatively low-risk but gives us a _breadth_ of testing internal testing can't.

### Stage 3 - Soft Release

At this stage, we consider the release to be "production ready" and ask will ask the community and our early testers to (partially) deploy the release to their production infrastructure.

**Goal(s):**

1. Test the release in some production environments with heavy workloads.
2. Partially roll-out an upgrade to see how it affects the network.
3. Retain the ability to ship last-minute fixes before the final release.

### Stage 4 - Release

At this stage, the release is "battle hardened" and ready for wide deployment.

## Performing a Release

> The first step is for the `Lead Maintainer` for `go-ipfs` to open an issue with Title `go-ipfs <version> Release` and a c&p of the [RELEASE_ISSUE_TEMPLATE](./RELEASE_ISSUE_TEMPLATE.md)

Author: ```suggestion
The release is managed by the `Lead Maintainer` for `go-ipfs`. It starts with the opening of an issue containing the content available on the [RELEASE_ISSUE_TEMPLATE](./RELEASE_ISSUE_TEMPLATE.md). Then, the 4 stages will be followed until the release is done.
```

# `go-ipfs` Release Flow

## Table of Contents

- [Release Philosophy](#release-philosophy)
- [Release Flow](#release-flow)
- [Performing a Release](#performing-a-release)
- [Release Version Numbers (aka semver)](#release-version-numbers-aka-semver)

## Release Philosophy

> `go-ipfs` aims to have release every six weeks, two releases per quarter. During these 6 week releases, we go through 4 different stages that gives us the opportunity to test the new version against our test environments (unit, interop, integration), QA in our current production environment, IPFS apps (e.g. Desktop and WebUI) and with our community and _early testers_<sup>[1]</sup> that have IPFS running in Production.

Reviewer 3: ```suggestion
`go-ipfs` aims to have release every six weeks, two releases per quarter. During these 6 week releases, we go through 4 different stages that gives us the opportunity to test the new version against our test environments (unit, interop, integration), QA in our current production environment, IPFS apps (e.g. Desktop and WebUI) and with our community and _early testers_<sup>[1]</sup> that have IPFS running in production.
```

2. Partially roll-out an upgrade to see how it affects the network.
3. Retain the ability to ship last-minute fixes before the final release.

### Stage 4 - Release

At this stage, the release is "battle hardened" and ready for wide deployment.

## Performing a Release

The first step is for the `Lead Maintainer` for `go-ipfs` to open an issue with Title `go-ipfs <version> Release` and a c&p of the [RELEASE_ISSUE_TEMPLATE](./RELEASE_ISSUE_TEMPLATE.md)

## Release Version Numbers (aka semver)

Until `go-ipfs` 0.4.X, `go-ipfs` was not using semver to communicate the type of release

Post `go-ipfs` 0.5.X, `go-ipfs` will use semver. This means that patch releases will not contain any breaking changes nor new features. Minor releases might contain breaking changes and always contain some new feature

Post `go-ipfs` 1.X.X (future), `go-ipfs` will use semver. This means that only major releases will contain breaking changes, minors will be reserved for new features and patches for bug fixes.

We do not yet retroactively apply fixes to older releases (no Long Term Support releases for now), which means that we always recommend users to update to the latest, whenever possible.

> TBW

Reviewer 3: Anything left here?

> TBW

Author: ```suggestion
```
